src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultClaimUidExtractor.cs

```diff
@@ -22,42 +22,94 @@ public DefaultClaimUidExtractor(ObjectPool<AntiforgerySerializationContext> pool
-        internal static IEnumerable<string> GetUniqueIdentifierParameters(ClaimsIdentity claimsIdentity)
+        public static IList<string> GetUniqueIdentifierParameters(IEnumerable<ClaimsIdentity> claimsIdentities)
         {
-            var nameIdentifierClaim = claimsIdentity.FindFirst(
-                claim => string.Equals(ClaimTypes.NameIdentifier, claim.Type, StringComparison.Ordinal));
-            if (nameIdentifierClaim != null && !string.IsNullOrEmpty(nameIdentifierClaim.Value))
+            if (claimsIdentities == null)
             {
-                return new string[]
+                return null;
+            }
+
+            foreach (var identity in claimsIdentities)
+            {
+                if (!identity.IsAuthenticated)
+                {
+                    continue;
+                }
+
+                var subClaim = identity.FindFirst(
+                    claim => string.Equals("sub", claim.Type, StringComparison.Ordinal));
+                if (subClaim != null && !string.IsNullOrEmpty(subClaim.Value))
+                {
+                    return new string[]
+                    {
+                        subClaim.Type,
+                        subClaim.Value,
+                        subClaim.Issuer
+                    };
+                }
+
+                var nameIdentifierClaim = identity.FindFirst(
+                    claim => string.Equals(ClaimTypes.NameIdentifier, claim.Type, StringComparison.Ordinal));
+                if (nameIdentifierClaim != null && !string.IsNullOrEmpty(nameIdentifierClaim.Value))
                 {
-                    ClaimTypes.NameIdentifier,
-                    nameIdentifierClaim.Value
-                };
+                    return new string[]
+                    {
+                        nameIdentifierClaim.Type,
+                        nameIdentifierClaim.Value,
+                        nameIdentifierClaim.Issuer
+                    };
+                }
+
+                var upnClaim = identity.FindFirst(
+                    claim => string.Equals(ClaimTypes.Upn, claim.Type, StringComparison.Ordinal));
+                if (upnClaim != null && !string.IsNullOrEmpty(upnClaim.Value))
+                {
+                    return new string[]
+                    {
+                        upnClaim.Type,
+                        upnClaim.Value,
+                        upnClaim.Issuer
+                    };
+                }
+            }
+
+            // We do not understand any of the ClaimsIdentity instances, fallback on serializing all claims in every claims Identity.
+            var allClaims = new List<Claim>();
+            foreach (var identity in claimsIdentities)
+            {
+                if (identity.IsAuthenticated)
+                {
+                    allClaims.AddRange(identity.Claims);
+                }
+            }
+
+            if (allClaims.Count == 0)
+            {
+                // No authenticated identities found.
```

Comment:
Yes. But we were checking of `IsAuthenticated` multiple times even before this change. Maybe for cases when someone uses a custom `IAntiforgeryTokenGenerator`.